### PR TITLE
[Hold] radicle-httpd: basic blob streaming support

### PR DESCRIPTION
**Pull request by keepsimple1**
This is to implement a basic blob streaming support mentioned in [this discussion](https://github.com/radicle-dev/heartwood/pull/331#discussion_r1110054271).

As `radicle-surf` does not concern itself with IO, nor async, I think it is better to support streaming in `heartwood` for blobs. @xphoniex .

Update:  after rebasing to #331 , I've updated `file_handler` to show the use of the streaming support.  See `BlobBody` in the patch.

### CODE REVIEW

File: radicle-httpd/src/api/json.rs [hunk 1]
```diff
@@ -64,6 +71,53 @@ pub(crate) fn blob<T: AsRef<[u8]>>(blob: &Blob<T>, path: &str) -> Value {
...
+#[allow(dead_code)]
+pub(crate) fn blob_stream<P: AsRef<Path>>(
+    blob: &[u8],
+    tmp_path: P,
+    chunk_sz: usize,
+) -> Result<FileStream, crate::Error> {
+    let tmp_path = tmp_path.as_ref();
+    let mut write_f = File::create(tmp_path)?;
+    write_f.write_all(blob)?;
+    FileStream::new(tmp_path, chunk_sz)
```
[the remark this answers is not in the document]
Comment (xphoniex) [reply radicle-httpd/src/api/json.rs:84]: (this applies to other fns/impls as well)

File: radicle-httpd/src/api/v1/projects.rs [hunk 2]
```diff
@@ -1310,4 +1310,34 @@ mod routes {
...
         );
     }
+
+    #[tokio::test]
+    async fn test_blob_stream() {
+        use crate::api::json::blob_stream;
+        use futures_util::StreamExt;
+
+        let tmp_dir = tempfile::tempdir().unwrap();
+        std::fs::create_dir_all(tmp_dir.path()).unwrap();
```
Comment (xphoniex) [inline radicle-httpd/src/api/v1/projects.rs:1320]: is this needed? doesn't `tempfile::tempdir()` already create the dir for you?

File: radicle-httpd/src/api/v1/projects.rs [hunk 3]
```diff
@@ -1310,4 +1310,34 @@ mod routes {
...
+        let tmp_file = tmp_dir.path().join("test_blob_stream");
+
+        let blob = "This is a test blob"; // 19 chars.
```
Comment (xphoniex) [inline radicle-httpd/src/api/v1/projects.rs:1323]: is comment needed?

File: radicle-httpd/src/api/v1/projects.rs [hunk 4]
```diff
@@ -1310,4 +1310,34 @@ mod routes {
             )
         );
     }
+
+    #[tokio::test]
+    async fn test_blob_stream() {
+        use crate::api::json::blob_stream;
+        use futures_util::StreamExt;
```
Comment (xphoniex) [inline radicle-httpd/src/api/v1/projects.rs:1317]: can you move these `use` statements outside?

**Review by xphoniex — COMMENTED** [review]
also prefix the branch name with your username please:
```
blob-stream` -> `keepsimple1/blob-stream
```

File: radicle-httpd/src/api/json.rs [hunk 5]
```diff
@@ -64,6 +71,53 @@ pub(crate) fn blob<T: AsRef<[u8]>>(blob: &Blob<T>, path: &str) -> Value {
     response
 }
 
+/// Creates a [`FileStream`] from `blob`, using `tmp_path` as the file.
+#[allow(dead_code)]
```
Comment (FintanH) [inline radicle-httpd/src/api/json.rs:75]: Why is this needed? :thinking:

File: radicle-httpd/src/api/json.rs [hunk 6]
```diff
@@ -64,6 +71,53 @@ pub(crate) fn blob<T: AsRef<[u8]>>(blob: &Blob<T>, path: &str) -> Value {
...
+    let mut write_f = File::create(tmp_path)?;
+    write_f.write_all(blob)?;
+    FileStream::new(tmp_path, chunk_sz)
+}
+
+/// Represents a read stream of a file.
+pub(crate) struct FileStream {
+    f: File,
+    chunk_sz: usize,
+}
```
Comment (FintanH) [inline radicle-httpd/src/api/json.rs:91]: I would have expected that you could have made a stream of the byte slice. Why are we opting to use a `File` here?

I'm pretty sure it's something along the lines of:
```rust
use futures::stream;

pub fn blob_stream(bs: &[u8]) -> impl Stream {
    stream::iter(bs)
}
```

The caller of `blob` can decide on their own chunking by calling [`chunks`](https://docs.rs/futures-util/0.3.26/futures_util/stream/trait.StreamExt.html#method.chunks).

File: radicle-httpd/src/api/v1/projects.rs (hunk 2 above)
Comment (FintanH) [reply radicle-httpd/src/api/v1/projects.rs:1320]: You need to call `path` to get the `&Path` out. `tempdir` gives you back a type `TmpDir`.

**Comment by keepsimple1** [comment]
> also prefix the branch name with your username please:
> 
> ```
> blob-stream` -> `keepsimple1/blob-stream
> ```

will do this in future.  (AFAIK, GitHub PR doesn't allow to change the source branch, unless creating a new PR).

File: radicle-httpd/Cargo.toml [hunk 7]
```diff
@@ -47,3 +48,4 @@ radicle-cli = { path = "../radicle-cli" }
 radicle-crypto = { path = "../radicle-crypto" }
 tempfile = { version = "3.3.0" }
 tower = { version = "0.4", features = ["util"] }
+futures-util = { version = "0.3" }
```
[the remark this answers is not in the document]
Comment (keepsimple1) [reply radicle-httpd/Cargo.toml:51]: will do.

[the remark this answers is not in the document]
Comment (keepsimple1) [reply radicle-httpd/Cargo.toml:51]: Btw I thought `cargo fmt` would take care of it, but I was wrong ;-)

File: radicle-httpd/src/api/json.rs [hunk 8]
```diff
@@ -1,5 +1,12 @@
 //! Utilities for building JSON responses of our API.
 
+use core::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use futures_core::Stream;
```
[the remark this answers is not in the document]
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs:7]: will do.

File: radicle-httpd/src/api/json.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs:84]: sure.

File: radicle-httpd/src/api/json.rs [hunk 9]
```diff
@@ -64,6 +71,53 @@ pub(crate) fn blob<T: AsRef<[u8]>>(blob: &Blob<T>, path: &str) -> Value {
...
+) -> Result<FileStream, crate::Error> {
+    let tmp_path = tmp_path.as_ref();
+    let mut write_f = File::create(tmp_path)?;
+    write_f.write_all(blob)?;
+    FileStream::new(tmp_path, chunk_sz)
+}
+
+/// Represents a read stream of a file.
+pub(crate) struct FileStream {
+    f: File,
```
[the remark this answers is not in the document]
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs:89]: no reasons ;-).  My "bad" habits from C programming ;-)

File: radicle-httpd/src/api/v1/projects.rs (hunk 3 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/v1/projects.rs:1323]: It's helpful for me to read my own test code ;-)

File: radicle-httpd/Cargo.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (xphoniex) [reply radicle-httpd/Cargo.toml:51]: there's [`cargo-sort`](https://crates.io/crates/cargo-sort) but I haven't used it yet.

File: radicle-httpd/src/api/v1/projects.rs (hunk 4 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/v1/projects.rs:1317]: will do.

File: radicle-httpd/src/api/json.rs (hunk 5 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs:75]: I haven't added a caller using this method, as I'm not familiar with how `httpd` wants to stream big files yet. Only the test case uses this in this PR.

File: radicle-httpd/src/api/json.rs (hunk 6 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs:91]: The essence of this patch is to support streaming from a file. The [motivation](https://github.com/radicle-dev/heartwood/pull/331#discussion_r1110054271) was to store a big blobs into a tmp file so that we can reduce memory consumption. Then `httpd` would stream the blob from the tmp file.

File: radicle-httpd/src/api/v1/projects.rs (hunk 2 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/v1/projects.rs:1320]: I will double check.

File: radicle-httpd/src/api/json.rs (hunk 5 above)
Comment (xphoniex) [reply radicle-httpd/src/api/json.rs:75]: you can re-base on https://github.com/radicle-dev/heartwood/pull/331 and show how it works in practice.

File: radicle-httpd/src/api/json.rs (hunk 6 above)
Comment (FintanH) [reply radicle-httpd/src/api/json.rs:91]: I see. It _seems_ a bit weird that you have to have the bytes in memory to avoid them being in memory -- but perhaps it makes sense.

Regardless, I don't think we need to roll our own here. I believe you could use a combination of [`futures::io::BufReader`](https://docs.rs/futures/0.3.26/futures/io/struct.BufReader.html), [`futures::io::AsyncBufReadExt::lines`](https://docs.rs/futures/0.3.26/futures/io/trait.AsyncBufReadExt.html#method.lines), and [`futures::StreamExt::chunks`](https://docs.rs/futures-util/0.3.26/futures_util/stream/trait.StreamExt.html#method.chunks) to achieve this. We would create the file and pass it to `BufReader::new`.

Also, because you're using a temporary directory, be _very_ careful that you hold onto the `TempDir` value until the stream is finished, otherwise it will be dropped and the directory will be removed.

File: radicle-httpd/src/api/v1/projects.rs (hunk 2 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/v1/projects.rs:1320]: You are correct. No need to create the dir (again). I added it as for some reason the dir was not created. Maybe there were some accidental mistakes I didn't notice. Now it's good 👍 . Thanks.

File: radicle-httpd/src/api/json.rs (hunk 6 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs:91]: > Also, because you're using a temporary directory, be _very_ careful that you hold onto the `TempDir` value until the stream is finished, otherwise it will be dropped and the directory will be removed.

yeah, I think this line holds a reference to the underlying `TempDir` in the test: 

```
let tmp_file = tmp_dir.path().join("test_blob_stream");
```

File: radicle-httpd/src/api/json.rs (hunk 5 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs]: I've rebased on #331 and changed the target branch to `dave/add-raw-route`.

Comment (xphoniex) [reply radicle-httpd/src/api/json.rs]: you haven't shown how this will be used in practice, inside [`file_handler`](https://github.com/radicle-dev/heartwood/commit/34b15f9f16d95de3f78e32ca4610280fdf16eb6c#diff-31099b31af3e5bf7f46f6dc4f2db59fdf633d2e65a3a06f48bfedbff35a29c9cR24).

Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs]: I think I did it [here](https://github.com/radicle-dev/heartwood/blob/446299ef832dce74500e3d5d0871e3efcf41f71d/radicle-httpd/src/raw.rs#L42-L43), no?

Comment (xphoniex) [reply radicle-httpd/src/api/json.rs]: Hm.. this is not how I was envisioning it. As Fintan mentioned:
> It seems a bit weird that you have to have the bytes in memory to avoid them being in memory -- but perhaps it makes sense.

instead of creating a tmp file, you can write it to a fixed path, e.g. `/tmp/<rid>/<hash>/*path` and inside `file_handler`, first check for existence of the file before loading the blob.

- file exists? stream from disk.
- file doesn't exist? 
  - file greater than `BLOB_STREAM_MIN_BYTES` ? save to disk & then stream
  - file is small ? return directly
  
this way, we only create the file once, and if 5 different people try to download a 200MB binary, server only has allocated 200MB once, for a short amount of time. This however, leaves server open to exhausting disk space. So we need a manager for these caches, a simple FIFO cache with a pre-determined space (e.g. 5 GB) would be ok here.

Comment (xphoniex) [reply radicle-httpd/src/api/json.rs]: On a second thought, an LRU cache is more appropriate here.

**Comment by cloudhead** [comment]
I don't understand the purpose of this PR.. We're reading the file from disk, then copying it back to disk, then streaming it from disk? The file is already in memory, so we should stream it from memory, or just not handle files larger than it makes sense to copy.

**Comment by xphoniex** [comment]
> 

By the time we know how big the file is, we have it in-memory. See my suggestion [here](https://github.com/radicle-dev/heartwood/pull/347#discussion_r1116587477).

**Comment by cloudhead** [comment]
Having it in-memory is not the only issue: bandwidth is another problem, as that's what will usually cost money. I don't think implementing any form of caching makes sense: the file is already on disk as a blob, we shouldn't have to create a copy of everything we want to serve.

**Comment by FintanH** [comment]
> I can add a size method on File [here](https://github.com/radicle-dev/radicle-git/blob/main/radicle-surf/src/fs.rs#L84), cause we can't use FileContent.

There's already a [`size`](https://github.com/radicle-dev/radicle-git/blob/main/radicle-surf/src/fs.rs#L148-L151) method on `FileContent`, so you can call `content` followed by `size` to get it.

To stream straight from disk, i.e. the git repository, we would probably need to bypass `git2` I would say and access the file system directly.

**Comment by cloudhead** [comment]
Yeah, I wouldn't bother bypassing git2. Simple solution is to just return an error if the file is too large, and otherwise copy it as we currently do.

File: radicle-httpd/src/api/json.rs (hunk 5 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs]: In that case, maybe we can go further and read / stream directly from the object file in git tree ?

Comment (xphoniex) [reply radicle-httpd/src/api/json.rs]: that seems to have its own complications, you can put this PR on hold for now.

**Comment by keepsimple1** [comment]
If we want to continue to explore, there are a couple of things we could try: 

1) stream the object file to the client, and let the client to process (i.e. decompress, remove its header).  
   - Pros: read directly from the disk, zero copy, minimum bandwidth
   - Cons: extra work in the client (front-end) code.

or 

2) store `repo` reference in the context / profile, so that `file_handler` can return a reference `&[u8]`. I also mentioned it [here ](https://github.com/radicle-dev/heartwood/pull/331#discussion_r1108759193) early.
   - Pros: read into memory once via `git2`. No additional copy or write to disk in our code.
   - Cons: Need to check: I hope `git2` won't copy the same object multiple times for multiple requests?  Will storing `repo` reference in the profile complicate other parts of `httpd`?

File: radicle-httpd/src/api/json.rs (hunk 5 above)
Comment (keepsimple1) [reply radicle-httpd/src/api/json.rs]: Is there a way to set the PR state to `hold` officially?

Comment (FintanH) [reply radicle-httpd/src/api/json.rs]: > Is there a way to set the PR state to hold officially?

You can turn it back into a draft

**Comment by keepsimple1** [comment]
As [discussed on zulip](https://radicle.zulipchat.com/#narrow/stream/369277-heartwood/topic/.E2.9C.94.20follow.20up.20on.20PR.20.23347/near/345026804),  I'm closing this PR for now.  Thanks!
